app/client/grpc.go

```diff
@@ -0,0 +1,149 @@
+package client
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"x-gwi/app/logs"
+	"x-gwi/app/pki"
+)
+
+const (
+	timeoutDial = 10 * time.Second
+)
+
+type GRPC struct {
+	config         *ConfigClient
+	grpcClientConn *grpc.ClientConn
+	cancel         context.CancelFunc
+	grpcDialOpts   []grpc.DialOption
+	mu             sync.RWMutex
+}
```

Comment:
There is a downside with using this approach - by using shared state struct
where you initialize it with New(), and then before using this, you always have to
call ValidConnection() (or DialContext()). There are however ways to do this differently.
Can you describe a different way to implement this? Can you think of a type-safe way to do it?

Reply:
btw there is experimental func (cc *ClientConn) GetState() connectivity.State
but it is stil experimental, also using mutex and is  more complex for the purpose

Reply:
grpc client and unused code is fixed now and this structure is not in use

If we talk about shared state structures vs mutable data structures vs immutable data structures it all depends on intention
I don't see downsides but rather features, consequences and requirements related to each approach
In the app I focus on zero-copy approach which provides great benefits in performance and of course require careful construction of the code with double-checking if there is no place changing shared value in unintentional way

If there is required immutable approach then we can construct without pointer as
type GRPC struct {
	config         ConfigClient

then use known patterns using getters, setters and withers like:
func (g GRPC) WithConfig(config   ConfigClient) GRPC {
    g.config = config
    return g
func (g GRPC) Config() ConfigClient {
    return g.config